module/zfs/arc.c

```diff
@@ -4043,10 +4043,11 @@ arc_evict_state(arc_state_t *state, uint64_t spa, int64_t bytes,
 
 		/*
 		 * Try to reduce pinned dnodes with a floor of arc_dnode_limit.
-		 * Request that 10% of the LRUs be scanned by the superblock
-		 * shrinker.
+		 * Request that zfs_arc_dnode_reduce_percent of the LRUs be
+		 * scanned by the superblock shrinker.
 		 */
-		if (type == ARC_BUFC_DATA && aggsum_compare(&astat_dnode_size,
+		if (type == ARC_BUFC_METADATA &&
```

Comment:
The existing code does appear to match the intent of the original commit 25458cbe (which includes a test case describing the issue).  But I wonder if the right solution here would instead be to do it for all `types` when over the limit and to wait for the task to complete.

Reply:
> ...and to wait for the task to complete.

See Also: https://github.com/openzfs/zfs/issues/7559#issuecomment-623795177

> Not related to this fix, but perhaps to the overall issue [i.e."arc_reclaim pegs CPU and blocks I/O"] : calling an async freeing function in a loop whilst trying to free up a specified amount of space seems problematic - it seems like we could fire off a whole bunch bunch of arc_prune_async() tasks whilst we're looping through "not enough free yet, let's try harder, not enough free yet, let's try harder, not enough free yet, let's try harder..." before the first async task has been able to even start up let alone do anything. I wonder if we should be using taskq_wait() or similar in there somewhere.

Reply:
I've updated the code with these suggestions. The thing that I'm not sure about is implication of the added `taskq_wait()` on the code paths that call `arc_evict_state()`

Reply:
**BIG RED WARNING FLAG**: I also have no idea if calling `taskq_wait()` in this context is safe. Sorry, I should have made that clear.